contracts/vesting/src/vesting.rs

```diff
@@ -0,0 +1,559 @@
+        // Returns the index of the next vesting period. Unless the current time is somehow in the past or vesting has not started yet.
+        // In case vesting is over it will always return NUM_VESTING_PERIODS.
+        match self
+            .periods
+            .iter()
+            .map(|period| period.start_time)
+            .collect::<Vec<u64>>()
+            .binary_search(&block_time.seconds())
+        {
+            Ok(u) => u,
+            Err(u) => u,
+        }
+    }
+
+    fn get_current_vesting_period(&self, block_time: Timestamp) -> usize {
+        if self.get_next_vesting_period(block_time) > 0 {
+            self.get_next_vesting_period(block_time) - 1
+        } else {
+            0
+        }
+    }
+
+    fn get_delegated_with_op(
+        &self,
+        op: &dyn Fn(u64, u64) -> bool,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        let block_time = block_time.unwrap_or(env.block.time);
+        let period = self.get_current_vesting_period(block_time);
+        if period == 0 {
+            return Coin {
+                amount: Uint128(0),
+                denom: DENOM.to_string(),
+            };
+        }
+
+        let end_time = if period >= NUM_VESTING_PERIODS as usize {
+            u64::MAX
+        } else {
+            self.periods[period].end_time().seconds()
+        };
+
+        if let Some(delegations) = get_account_delegations(storage, &self.address) {
+            delegations
+                .iter()
+                .filter(|d| op(d.block_time.seconds(), end_time))
+                .fold(
+                    Coin {
+                        amount: Uint128(0),
+                        denom: DENOM.to_string(),
+                    },
+                    |mut acc, d| {
+                        acc.amount += d.amount;
+                        acc
+                    },
+                )
+        } else {
+            Coin {
+                amount: Uint128(0),
+                denom: DENOM.to_string(),
+            }
+        }
+    }
+
+    fn get_balance(&self, storage: &dyn Storage) -> Uint128 {
+        get_account_balance(storage, &self.address).unwrap_or(Uint128(0))
+    }
+}
+
+impl DelegationAccount for PeriodicVestingAccount {
+    fn try_delegate_to_mixnode(
+        &self,
+        mix_identity: IdentityKey,
+        coin: Coin,
+        env: &Env,
+        storage: &mut dyn Storage,
+        querier: Option<QuerierWrapper>,
+    ) -> Result<(), ContractError> {
+        if coin.amount < self.get_balance(storage) {
+            if let Some(querier) = querier {
+                let msg = MixnetExecuteMsg::DelegateToMixnodeOnBehalf {
+                    mix_identity: mix_identity.clone(),
+                    delegate_addr: self.address.clone(),
+                    coin: coin.clone(),
+                };
+                querier.query_wasm_smart(DEFAULT_MIXNET_CONTRACT_ADDRESS, &msg)?;
```

Comment:
This is more of a question than a comment: even though it's called `query_...`, it can still do `execute` calls, not only queries, right?

Reply:
Thats an excellent question, and I'm not really sure, the docs don't say exactly, I'll have to run that by the `cosmwasm` folks

Reply:
I have a thread there, there is also `wasm_execute` that is probably a better candidate for the job here...

Reply:
I think so, yeah. This can't really work as you're not sending funds anywhere for the delegation to happen. And in order to send those funds, there is going to be a hard problem of where do these funds come from. From the vested or the unvested tokens? What happens when you enter a new period and other tokens become liquid as well?

Reply:
vested vs unvested is not important here, everything can be delegated, [this](https://github.com/nymtech/nym/tree/vesting-staking-cointract/contracts/vesting#vesting-coin-delegation-flow) is how the flow should go

Reply:
But how do you know from what pool to send tokens do be delegated. In the current PR, I don't see there any funds being sent on `DelegateToMixnodeOnBehalf`, but I don't think that call would work, as the mixnet contract expects the funds that will be delegated to be specified in the execute message.

Reply:
[Here](https://github.com/nymtech/nym/blob/vesting-staking-cointract/contracts/vesting/src/vesting.rs#L243) is where we add funds to the message and then return it as a `Response`, which as far as I understand should send the message out